Add syslog supports

tisonkun opened the pull request:
This refers to https://github.com/nix-rust/nix/issues/413.

Hello world works locally:

```
2024-11-12 19:19:52.264557+0800 0x27405d0  Default     0x0                  27738  0    test-aae182a0654f64e3: Hello, nix!
```

I'd like to send this patch early for any necessary alignments.

SteveLauC commented:
BTW, a changelog is needed, please see [CONTRIBUTING.md](https://github.com/nix-rust/nix/blob/master/CONTRIBUTING.md#pull-requests) on how to add one.

tisonkun commented:
@SteveLauC Thanks for your information! All updated.

> For enum, [libc_enum!](https://github.com/nix-rust/nix/blob/0e4353a368abfcedea4ebe4345cf7604bb61d238/src/macros.rs#L101) will be used. Types Severity and Facility fall into this category.

Updated.

> If a C string is needed (string with a tailing NUL), [trait NixPath](https://docs.rs/nix/latest/nix/trait.NixPath.html) can be used to fill the gap. You can use it for the void openlog(const char *ident argument.

Updated.

> The priority argument of syslog() and vsyslog() should be constructed by ORing enum Severity and enum Facility, we can probably impl std::ops::BitOr<Output = c_int> for these 2 types, but this is not that fault-resilient, perhaps we should define another wrapper type here🤔

Reasonable. `Priority` struct added.

> Regarding the format void syslog(int priority, const char *format, ...); argument, do you think we should implement syslog as a macro similar to println!?

No. As stated at https://github.com/nix-rust/nix/issues/413#issuecomment-2470194025, I think we just simply use Rust's formatting function and pass the formatted string. This is also how Python wraps syslog functionalities.

> A few tests to ensure the wrapper would work should be sufficient.

Included.

tisonkun commented:
This patch doesn't wrap `vsyslog` and `setlogmark`.

For `vsyslog`, it's a variant of `syslog` for ergonomically coding in C, irrelevant to Rust.

For `setlogmark`, `libc` doesn't export `LOG_MASK` and/or `LOG_UPTO`, make it hard to wrap reasonable code. I temporarily added the wrapper but later revert it. IMO it's less used and we can revisit this function when there is a real-world necessity.

SteveLauC commented on src/syslog.rs:
```diff
@@ -0,0 +1,146 @@
+/// argument specifies logging options. The `facility` parameter encodes a default facility to be
+/// assigned to all messages that do not have an explicit facility encoded.
+pub fn openlog<P: NixPath + ?Sized>(
+    ident: &P,
```
On Linux, the `ident` argument is suggested to have the static lifetime: 

> The argument ident in the call of `openlog()` is probably stored as-is.  Thus, if the string it points to is changed, `syslog()` may start prepending the changed string, and if the string it points to ceases to exist, the results are undefined.  Most portable is to use a string constant.

I think we will have a different interface for Linux in the future:

```rs
pub fn openlog(
    ident: &'static Cstr,
```

SteveLauC replied:
I just realized that using `S: AsRef<OsStr> + ?Sized>` looks exactly right for the `syslog(..., message)` argument, maybe `ident` should also be done similarly, considering they are not paths but strings.

SteveLauC commented on src/syslog.rs:
```diff
@@ -0,0 +1,146 @@
+///
+/// The message is then written to the system console, log files, logged-in users, or forwarded
+/// to other machines as appropriate.
+pub fn syslog<S: AsRef<OsStr> + ?Sized>(
```
Can we have an example here? It would be great if it could show how to use Rust's built-in formatter with this function.

tisonkun commented:
`LOG_MASK` and `LOG_UPTO` may be easy to add:

```rust
#[macro_export]
macro_rules! LOG_MASK 
{
    ($($arg:tt)*) => (
        (1 << $($arg)*)
    )
}

#[macro_export]
macro_rules! LOG_UPTO 
{
    ($($arg:tt)*) => (
        ((1 << (($($arg)*) + 1)) - 1)
    )
}
```

But I'd like to wait for this patch getting review and perhaps did it in a followup so that we don't push everything at once.

SteveLauC commented on src/syslog.rs:
```diff
@@ -0,0 +1,146 @@
+    priority: Priority,
+    message: &S,
+) -> Result<()> {
+    let formatter = OsStr::new("%s");
```
Neat! This also conforms to a suggestion in the Linux man page:

> Never pass a string with user-supplied data as a format, use the following instead:
>
> ```c
> syslog(priority, "%s", string);
> ```

SteveLauC commented on changelog/2537.added.md:
```diff
@@ -0,0 +1 @@
+Add support for `syslog`, `openlog`, `closelog` on `macos`.
```
Looks like `closelog()` is not included in the PR

tisonkun replied:
Weird .. Perhaps get reverted in the setlogmark func. Let me add it back.

SteveLauC commented on src/syslog.rs:
```diff
@@ -0,0 +1,146 @@
+    /// Severity levels for log messages.
+    #[repr(i32)]
+    #[non_exhaustive]
+    pub enum Severity {
```
The man pages of macOS, FreeBSD, and Linux all use the word "level", what about renaming this type to simply `LogLevel`?

tisonkun replied:
Severity is defined as in:

* https://en.wikipedia.org/wiki/Syslog
* https://datatracker.ietf.org/doc/html/rfc3164
* https://datatracker.ietf.org/doc/html/rfc3164

I'd prefer to use this name.

SteveLauC replied:
Makes sense

SteveLauC commented on src/syslog.rs:
```diff
@@ -0,0 +1,146 @@
+
+libc_bitflags! {
+    /// Options for system logging.
+    pub struct LogFlags: libc::c_int {
```
Thanks for following the [convention](https://github.com/nix-rust/nix/issues/317)!

tisonkun replied:
I didn't even notice this. Just find a name I feel comfortable, lol.

tisonkun commented:
Just noticed that we can't use `with_nix_path` here because NixPath has a limit on length which seems not applicable for ident and message.

tisonkun commented:
@SteveLauC test on unix and updated.

I have an idea to implement `setlogmask`, but not sure if we can export `libc::c_int` or we have other better types:

```rust
pub fn setlogmask(maskpri: libc::c_int) -> libc::c_int {
    let mask = unsafe { libc::setlogmask(maskpri) };
    mask
}

pub fn log_mask(pri: Severity) -> libc::c_int {
    1 << (pri as libc::c_int)
}

pub fn log_upto(pri: Severity) -> libc::c_int {
    (1 << ((pri as libc::c_int) + 1)) - 1
}
```

tisonkun commented on src/lib.rs:
```diff
@@ -200,6 +202,12 @@ feature! {
     pub mod spawn;
 }
 
+#[cfg(unix)]
```
I've tested on ubuntu and archlinux also.

Perviously add `cfg(any(macos, linux))` and clippy notice me to make it `cfg(unix)`.

tisonkun commented on test/test.rs:
```diff
@@ -42,6 +42,9 @@ mod test_sendfile;
 ))]
 mod test_spawn;
 
+#[cfg(unix)]
```
Or may be just removed it? Since nix works only for UNIX already.

SteveLauC commented on src/syslog.rs:
```diff
@@ -0,0 +1,193 @@
+use crate::Result;
+use std::ffi::OsStr;
+
+/// Logging options of subsequent [`syslog`] calls can be set by calling [`openlog`].
+///
+/// The parameter `ident` is a string that will be prepended to every message. The `logopt`
+/// argument specifies logging options. The `facility` parameter encodes a default facility to be
+/// assigned to all messages that do not have an explicit facility encoded.
+pub fn openlog<S>(
+    ident: Option<&S>,
```
This interface should be good for UNIXs other than Linux, since [Linux would store the `ident` pointer as-is](https://github.com/nix-rust/nix/pull/2537#discussion_r1844889953) (will be used during whole process lifetime), we need to add a separate interface for Linux, and gate this with `#[cfg(not(target_os = "linux"))]`.

SteveLauC replied:
And thanks for adding the Option here! I didn't catch it in my first round of review.

tisonkun replied:
I'd suggest we remove the `ident` param for now.

It's not easy to just add `Option<&'static str>` because it may not end with `nul` and I don't know how to achieve this.

tisonkun replied:
I don't observe the `ident` on macos and linux, perhaps this is because the allocated string is collected 🤣

tisonkun replied:
Maybe `Option<&'static CStr>`. Let me try.

SteveLauC replied:
Nix does not pursue uniform interfaces across OSes, if they are different, let them be different. I suggest keeping the original interface for non-Linux platforms as they are not constrained by this limit and adding a separate interface for Linux

tisonkun replied:
Could you send a patch to this PR or just edit this PR with the suggested change?

SteveLauC replied:
> Could you send a patch to this PR

Yeah, let me send a commit

SteveLauC replied:
See https://github.com/nix-rust/nix/pull/2537/commits/596280b34a64018923bf9eb85e3ec7fb679ced55

AlexTMjugador replied:
I just stumbled upon this review comment when writing a portable interface that relies on `openlog` and I found the different signature among different Unix flavors quite strange, and potentially inappropriate.

`openlog` is a function provided by the environment C library, which may be implemented in different ways even for the same Unix flavor (Linux targets routinely use both `musl` and `glibc` as C library implementations), so changing a function signature just because the target is Linux is something that should be done with lots of caution, as it's likely to not be proper.

Moreover, the [POSIX standard](https://pubs.opengroup.org/onlinepubs/9799919799/functions/syslog.html) says that "the application shall ensure that the string pointed to by `ident` remains valid during the `syslog()` calls that will prepend this identifier". Because we can't really know in this function at what times `syslog()` will be further called, the only safe lifetime for the `ident` parameter is `'static`. And because the string is assumed to be a standard C string with no NUL characters in the middle, it should be `CStr`. Thus, I think the Linux signature of this function is the one that should be used for all Unix flavors.

What do you think about this? Would a PR to tweak this be welcome? :smile:

SteveLauC commented on src/syslog.rs:
````diff
@@ -0,0 +1,193 @@
+/// ```
+pub fn syslog<P, S>(priority: P, message: &S) -> Result<()>
+where
+    P: Into<Priority>,
````
Nice!

SteveLauC commented on src/syslog.rs:
```diff
@@ -0,0 +1,178 @@
+        /// which file descriptors are allocated.
+        LOG_NDELAY;
+        /// Write the message to standard error output as well to the system log.
+        #[cfg(all(not(target_os = "redox"), not(target_os = "illumos")))]
```
Let's do:

```suggestion
        #[cfg(not(any(target_os = "redox", target_os = "illumos")))]
```

This is easier to read and feels more natural.

tisonkun commented:
> @SteveLauC test on unix and updated.
> 
> I have an idea to implement `setlogmask`, but not sure if we can export `libc::c_int` or we have other better types:
> 
> ```rust
> pub fn setlogmask(maskpri: libc::c_int) -> libc::c_int {
>     let mask = unsafe { libc::setlogmask(maskpri) };
>     mask
> }
> 
> pub fn log_mask(pri: Severity) -> libc::c_int {
>     1 << (pri as libc::c_int)
> }
> 
> pub fn log_upto(pri: Severity) -> libc::c_int {
>     (1 << ((pri as libc::c_int) + 1)) - 1
> }
> ```
> 
> Also related test ideas - https://stackoverflow.com/questions/56049852/python-2-7-setlogmask0-not-disabling-syslog

@SteveLauC Here is an open question on whether to include `setlogmask` in this PR.

SteveLauC commented:
Regarding the `setlogmask()` interface, I think we can implement it in the following way:

```rs
/// System log priority mask.
#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct LogMask(libc::c_int);

/// Set the process-wide priority mask to `mask` and return the previous mask 
/// value. 
/// 
/// Calls to `syslog()` with a priority level not set in `mask` are ignored. The 
/// default is to log all priorities.
/// 
/// If the `mask` argument is `None`, the current logmask is not modified, this
/// can be used to query the current log mask.
pub fn setlogmask(mask: Option<LogMask>) -> LogMask {
    let mask = match mask {
        Some(mask) => mask.0,
        None => 0,
    };
    
    let prev_mask = unsafe {
        libc::setlogmask(mask)
    };
    
    LogMask(prev_mask)
}

impl LogMask {
    /// Creates a mask of all priorities up to and including `priority`.
    #[doc(alias("LOG_UPTO"))]
    pub fn up_to(priority: Severity) -> Self {
        todo!()
    }
    
    /// Creates a mask for the specified priority.
    #[doc(alias("LOG_MASK"))]
    pub fn of_priority(priority: Severity)  -> Self {
        todo!()
    }
    
    /// Returns if the mask for the specified `priority` is set.
    pub fn contains(&self, priority: Severity) -> bool {
        todo!()
    }
}

impl std::ops::BitOr for LogMask {
    type Output = Self;
    fn bitor(self, rhs: Self) -> Self::Output {
        Self(self.0 | rhs.0)
    }
}
``` 

Though I think the C macros `LOG_MASK` and `LOG_UPTO` should be added to the libc crate first, so perhaps we can include these things in the next PR?

tisonkun commented:
> Though I think the C macros LOG_MASK and LOG_UPTO should be added to the libc crate first, so perhaps we can include these things in the next PR?

Yes. Let's review this PR as is now.

@SteveLauC Thanks for your following commits!

SteveLauC reviewed (approved):
Looks good, thanks for the PR, let's merge it! 🚀
